pep-0577.rst

```diff
@@ -351,32 +406,72 @@ type checkers would be encouraged to enforce the same "declaration or assignment
+
+For module global scopes, the answer to both of those questions is unequivocally
+"Yes", because it's a language level guarantee that mutating the ``globals()``
+dict will immediately impact the runtime module scope, and ``global NAME``
+declarations inside a function can have the same effect (as can importing the
+currently executing module and modifying its attributes).
+
+For class scopes, the answer to both questions is also "Yes" in practice,
+although less unequivocally so, since the semantics of ``locals()`` are
+currently formally unspecified. However, if the current behaviour of ``locals()``
+at class scope is taken as normative (as PEP 558 proposes), then this is
+essentially the same scenario as manipulating the module globals, just using
+``locals()`` instead.
+
+For function scopes, however, the current answers to these two questions are
+respectively "Yes" and "No". Expression level rebinding of function locals is
+already possible thanks to lexically nested scopes and explicit ``nonlocal NAME``
+expressions. While this PEP will likely make expression level rebinding more
+common than it is today, it isn't a fundamentally new concept for the language.
+
+By contrast, declaring a *new* function local variable is currently a statement
+level action, involving one of:
+
+* an assignment statement (``NAME = EXPR``, ``OTHER_TARGET = NAME = EXPR``, etc)
+* a variable declaration (``NAME : EXPR``)
+* a nested function definition
+* a nested class definition
+* a ``for`` loop
+* a ``with`` statement
+* an ``except`` clause (with limited scope of access)
+
+The historical trend for the language has actually been to *remove* support for
+expression level declarations of function local names, first with the
+introduction of "fast locals" semantics (which made the introduction of names
+via ``locals()`` unsupported for function scopes), and again with the hiding
+of comprehension iteration variables in Python 3.0.
+
+Now, it may be that in Python 3.9, we decide to revisit this question based on
+our experience with expression level name binding in Python 3.8, and decide that
+we really do want expression level function local variable declarations as well,
+and that we want ``NAME := EXPR`` to be the way we spell that (rather than,
+for example, spelling inline declarations more explicitly as
+``NAME := EXPR given NAME``, which would permit them to carry type annotations,
+and also permit them to declare new local variables in scoped expressions,
+rather than having to pollute the namespace in their containing scope).
+
+But the proposal in this PEP is that we explicitly give ourselves a full
+release to decide how much we want that feature, and exactly where we find
+its absence irritating. Python has survived happily without expression level
+name bindings *or* declarations for decades, so we can afford to give ourselves
+a couple of years to decide if we really want *both* of those, or if expression
+level bindings are sufficient.
```

Comment:
This is well reasoned.

Reply:
Noting that every example in my PEP 572 Appendix would blow up (none of which involved a scoped expression, BTW).  The majority of cases are simple and benefit from less redundancy and from removing a line of code.  All of that is lost if I need to, e.g., add a semantically senseless `diff = g = None` before

    if (diff := x - x_base) and (g := gcd(diff, n)) > 1:

or, possibly worse, give a type annotation I have no other use for (worse because then I _also_ need to add `import typing`).  Why is this code binding things to `None` when they're immediately re-bound?  Or declaring their types in code that otherwise doesn't use types at all?  Those questions can make code _more_ mysterious.

I don't particularly mind if I need to add something to establish the intended scope for targets in scoped expressions, but for block-level code the intent of `x := e` in an expression is just plain obvious.

Reply:
Aye, I'd noticed that too. One intermediate option would be to only require the pre-declaration in cases where an explicit nested non-local reference would currently cause a syntax error. That said, if `NAME: ...`  declared the scope of a name while still inferring its type from the first binding, then the required boilerplate would be relatively minimal.

I'll tinker with it a bit, since I still prefer having a real semantic reason for favouring the established assignment statements over assignment expressions in cases where the two collide.